### PR TITLE
Databases: add support for zip archiving

### LINKED ISSUE #87: Database: Zip folder archiving
Sometimes local folders outside of a database have to be archived. Being able to provide a local folder (that could eventually be bind mounted inside the Blackbox container) that could be zipped using the `zipfile` stdlib and backed up like any database dump would be quite useful. Compression level could also be added as an option.

I’d be interested in implementing this.

### PR DESCRIPTION
**Pull request by Akarys42**
Closes #87 

This commit adds support for recursively storing a local file as a Zip file.

### CODE REVIEW

File: blackbox/handlers/databases/zip.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    def backup(self) -> Path:
+        date = datetime.date.today().strftime("%d_%m_%Y")
+
+        path = self.config["path"]
+        compression_level = self.config.get("compression_level", 5)
+
+        if compression_level < 0 or compression_level > 9:
+            raise ImproperlyConfigured(
+                f"Invalid compression level. Must be an integer between 0 and 9, got {compression_level}."
+            )
```
Comment (lemonsaurus) [inline blackbox/handlers/databases/zip.py:25]: I think this is too late to raise this error. At this point, we might already have backed things up to several databases.

Maybe we can do this in the `__init__` for this model, instead? That way it should fail instantly, before taking any actions.

File: blackbox/handlers/databases/zip.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            directories = deque((Path(path),))
+
+            while not len(directories) == 0:
+                path = directories.pop()
+
+                for file in path.iterdir():
+                    if file.is_file():
+                        zipfile.write(file)
+                    elif file.is_dir():
+                        directories.append(file)
```
Comment (lemonsaurus) [inline blackbox/handlers/databases/zip.py:39]: This is super complex, and can be simplified a lot using something like `.rglob`, e.g.

```py
for subpath in path.rglob("*"):
    if subpath.is_file():
        zipfile.write(subpath)
```

**Review by lemonsaurus — CHANGES_REQUESTED** [review]
Thanks for the PR, Akarys!

A few problems:
- Handler is not well enough documented. Please add some block comments explaining what you're doing.
- I'm not sure about calling this storage type zip. Sure, we're zipping the backup, but we're _storing it_ locally. So I would argue that semantically, it makes more sense for the storage type to be something like "local storage". The fact that we zip it is an implementation detail.

File: blackbox/handlers/databases/zip.py (hunk 2 above)
Comment (Akarys42) [reply blackbox/handlers/databases/zip.py]: I didn't know that was a thing, that's such a more elegant method :grin:

File: README.md [hunk 3]
```diff
@@ -277,7 +277,7 @@ databases:
 
 ## Databases
 
-Right now, this app supports **MongoDB**, **PostgreSQL 13**, **MariaDB**, **Redis** and **Zip archiving**. If
+Right now, this app supports **MongoDB**, **PostgreSQL 13**, **MariaDB**, **Redis** and **LocalStorage archiving**. If
```
Comment (lemonsaurus) [inline README.md:280]: this feels weird, and may be confused with `localStorage`, the HTML5 feature.

Let's just make it **local storage archiving** without capital letters.

File: README.md [hunk 4]
```diff
@@ -371,14 +371,21 @@ Identifiers can be any string of your choosing.
       port: "6379"
 ```
 
-### Zip archiving
+### Local Storage
```
Comment (lemonsaurus) [inline README.md:374]: let's do `Local storage` here.

File: blackbox/handlers/databases/localstorage.py [hunk 5]
```diff
@@ -13,30 +12,33 @@ class LocalStorage(BlackboxDatabase):
...
-            while not len(directories) == 0:
-                path = directories.pop()
-
-                for file in path.iterdir():
-                    if file.is_file():
-                        zipfile.write(file)
-                    elif file.is_dir():
-                        directories.append(file)
+            for file in path.rglob("*"):
+                if file.is_file():
```
Comment (lemonsaurus) [inline blackbox/handlers/databases/localstorage.py:39]: this line irks me a bit. why did we name the variable `file` if we have to check to make sure it is a file? that's why my suggestion had `subpath`, but I'm open to other more "zoomed out" alternatives. I just don't think we should name something file and then check if it's a file.

**Review by lemonsaurus — CHANGES_REQUESTED** [review]
Getting closer.

**Review by lemonsaurus — APPROVED** [review]
Looks viable. Let's gooo. :shipit:

**Pull request merged**